### DISCUSSION TITLE
How to remove `# nolint`, `# styler: off` from being rendered out

### DISCUSSION THREAD

**Discussion by pawelru** [quarto-dev/quarto-cli#8760] [Q&A]
### Description

Suppose I have the following:

````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

```{r}
ABC <- 1 # nolint object_name_linter.
print(ABC)
```

````

How can I exclude `# nolint ...` from being rendered out?

I have found that knitr supports `tidy` chunk option of custom function ([source](https://yihui.org/knitr/options/#:~:text=reformat%20the%20code.-,A%20custom%20function,-of%20the%20form)) which in my case will look like this: `function(code, ...) gsub(pattern = "#\\s?nolint.*|#\\s?styler:.*", replacement = "", code)`.

This however seems to be not supported ([source](https://github.com/quarto-dev/quarto-cli/blob/c3419362033f56b1029aa1e5ff93871294192102/src/resources/schema/cell-codeoutput.yml#L117-L125)).

I can of course go ahead and add it in each chunk I have but that's not so DRY and I am interested in a _global_ setting (presumably specified in `_quarto.yaml` config file).

Initially I wanted to post it as a feature request but before that - do you see any other way (not necessarily via `tidy` option) to achieve what I want? 

Related: https://github.com/quarto-dev/quarto-cli/discussions/793

**Comment by mcanouil** [comment]
You can define a Lua filter, see <https://quarto.org/docs/extensions/filters.html>.

Knitr specific options needs to be passed according to the documentation: https://quarto.org/docs/computations/r.html#knitr-options

For more, please provide a small **self-contained "working" (reproducible)** example to work with, _i.e._, a complete Quarto document or a Git repository.

---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

![A placeholder image](https://placehold.co/600x400.png)

The end.
````
`````

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Reply by pawelru** [reply]
Thank you. I will have a look at the filters. Thank you for the reference.

I tried with knitr options. My config yaml file looked like this:
```
knitr:
  opts_chunk:
    echo: true
    tidy: function(code, ...) gsub(pattern = "#\\s?nolint.*|#\\s?styler:.*", replacement = "", code)
```

But this gave me error:
```
1: In eng_r(options) :
  Failed to tidy R code in chunk 'knitr_utils'. Reason:
Error in do.call(tidy.method, c(list(code), options$tidy.opts)) : 
  'what' must be a function or character string
```

That's probably because of Quarto accepts `tidy` parameter to be of either boolean or one of ["styler", "formatR"]. `knitr` however accepts also a function.

I also modified my initial posts with to make it more reproducible - hope it helps.

**Reply by mcanouil** [reply]
Still, you need to provide a full reproducible example as shown, not simply a snippet and an error message.
`echo` should be defined under `execute` key not within `knitr.opts_chunk`.

Note that you can still use the old way:

````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

```{r}
#| include: false
knitr::opts_chunk$set(
  tidy = function(code, ...) {
    gsub(pattern = "#\\s?nolint.*|#\\s?styler:.*", replacement = "", code)
  }
)
```

```{r}
ABC <- 1 # nolint object_name_linter.
print(ABC)
```
````

**Reply by pawelru** [reply]
> Still, you need to provide a full reproducible example as shown, not simply a snippet and an error message.

You mean that variant with `tidy` in yaml file? This is clearly not supported as per manual and codebase - I guess there is nothing to reproduce :)

> echo should be defined under execute key not within knitr.opts_chunk.

Apologies - my mistake. I will fix it separately. I assume this does not impact the main issue there.

> Note that you can still use the old way:

True but I'm dealing with ~100 articles and looking for a single, _global_ place to configure it. AFAIK I can define define that single code chunk in one file and use include shortcode in all articles to load it but still there would be ~ 100 include statements. If there is no other way I will probably go with this. The question is if there is a more DRY way.

**Reply by mcanouil** [reply]
What you showed from the codebase is the code cell option.

Although I don't understand the point to make that many linter exclusion on code cells.

You could simply define `.lintr` and exclude rules.

Here the issue is that YAML does not take R code.

The best and more general approach remains a LUA filter, especially because contrary to knitr it is not limited to R code cell.

Edit: I believe there is an example of such filter in here. (Cannot search for it now)

**Reply by mcanouil** [reply]
FYI:
- https://github.com/quarto-dev/quarto-cli/discussions/7521

**Reply by pawelru** [reply]
Thank you for all good hints. I went ahead and created a custom lua filter as you suggested.
For everyone else coming back to this in the future it's available here: https://github.com/insightsengineering/pattern-strip
